src/main/java/dev/morling/demos/logicaldecoding/Main.java

```diff
@@ -51,31 +57,26 @@ public static void main(String[] args) throws Exception {
                 .deserializer(new JsonDebeziumDeserializationSchema()) // converts SourceRecord to JSON String
                 .build();
 
-        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        // for local development, also start the Flink UI on http://localhost:8081
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(new Configuration());
+        env.setParallelism(2);
+        env.getCheckpointConfig().setCheckpointInterval(Time.seconds(5).toMilliseconds());
 
-        env.addSource(sourceFunction)
-                .keyBy(v -> 1L)
-                .map(new AuditMetadataEnrichmentFunction())
-                .print()
-                .setParallelism(1);
+        env.addSource(sourceFunction).forceNonParallel()
+                .map(new AuditMetadataEnrichmentFunction()).forceNonParallel()
+                .print();
 
         env.execute();
     }
 
-    public static class AuditMetadataEnrichmentFunction extends RichMapFunction<String, String> {
+    public static class AuditMetadataEnrichmentFunction extends RichMapFunction<String, String> implements CheckpointedFunction {
 
         private static final long serialVersionUID = 1L;
 
-        private final ObjectMapper mapper;
+        private transient ObjectMapper mapper;
 
-        private transient ValueState<AuditState> auditState;
-
-        public AuditMetadataEnrichmentFunction() {
-            mapper = new ObjectMapper();
```

Comment:
The mapper was not transient & initialized in the constructor. This means the ObjectMapper was part of the closure send from the main() method to the cluster as part of the UDF.
I'm now initializing the mapper in the open() method, to be part of the Flink execution lifecycle.